src/simulators/tensor_network/MPS.cpp

```diff
@@ -0,0 +1,398 @@
+/**
+ * Copyright 2018, IBM.
+ *
+ * This source code is licensed under the Apache License, Version 2.0 found in
+ * the LICENSE.txt file in the root directory of this source tree.
+ */
+
+
+#include <bitset>
+#include <math.h>
+
+// for analysis of memory consumption
+#include "stdlib.h"
+#include "stdio.h"
+#include "string.h"
+
+#include "framework/utils.hpp"
+
+#include "MPS.hpp"
+#include "MPS_tensor.hpp"
+
+namespace AER {
+namespace TensorNetworkState {
+
+uint_t reverse_bits(uint_t num, uint_t len);
+vector<uint_t> calc_new_indexes(vector<uint_t> indexes);
+
+template <class T>
+void myswap(T &a, T &b){
+	T temp = a;
+	a = b;
+	b = temp;
+}
+
+uint_t reverse_bits(uint_t num, uint_t len) {
+  uint_t sum = 0;
+  //  std::assert(num < pow(2, len));
+  for (uint_t i=0; i<len; ++i) {
+    if ((num & 0x1) == 1) {
+      sum += pow(2, len-1-i);
+    }
+    num = num>>1;
+    if (num == 0) {
+      break;
+    }
+  }
+  return sum;
+}
+
+vector<uint_t> calc_new_indexes(vector<uint_t> indexes)
+{
+	uint_t n = indexes.size();
+	uint_t avg = round(accumulate( indexes.begin(), indexes.end(), 0.0)/ n );
+	vector<uint_t> new_indexes( n );
+	std::iota( std::begin( new_indexes ), std::end( new_indexes ), avg-n/2);
+	return new_indexes;
+}
+
+void MPS::initialize(uint_t num_qubits)
```

Comment:
In principle, I agree with this comment as well as the next. However, I tried to be compatible with the other simulation methods. They all implement an empty constructor, and then they call initialize_qreg(), which calls qreg_.initialize(). Since initialize_qreg() is abstract in base class State, I must implement it. What do you think is best?